utils/script.c

```diff
@@ -0,0 +1,420 @@
+static PyAPI_FUNC(int) (*__PyTuple_SetItem)(PyObject *, Py_ssize_t, PyObject *);
+static PyAPI_FUNC(PyObject *) (*__PyObject_GetAttrString)(PyObject *, const char *);
+static PyAPI_FUNC(PyObject *) (*__PyObject_CallObject)(PyObject *callable_object, PyObject *args);
+
+static PyAPI_FUNC(PyObject *) (*__PyString_FromString)(const char *);
+static PyAPI_FUNC(PyObject *) (*__PyInt_FromLong)(long);
+static PyAPI_FUNC(PyObject *) (*__PyLong_FromUnsignedLongLong)(unsigned PY_LONG_LONG);
+
+static PyAPI_FUNC(char *) (*__PyString_AsString)(PyObject *);
+static PyAPI_FUNC(long) (*__PyLong_AsLong)(PyObject *);
+
+static PyAPI_FUNC(PyObject *) (*__PyDict_New)(void);
+static PyAPI_FUNC(int) (*__PyDict_SetItem)(PyObject *mp, PyObject *key, PyObject *item);
+static PyAPI_FUNC(int) (*__PyDict_SetItemString)(PyObject *dp, const char *key, PyObject *item);
+static PyAPI_FUNC(PyObject *) (*__PyDict_GetItem)(PyObject *mp, PyObject *key);
+
+static PyAPI_FUNC(PyObject *) (*__PyLong_FromLong)(long);
+
+static PyObject *pName, *pModule, *pFuncEntry, *pFuncExit;
+
+extern struct symtabs symtabs;
+
+enum py_args {
+	PY_ARG_TID = 0,
+	PY_ARG_DEPTH,
+	PY_ARG_START_TIME,
+	PY_ARG_END_TIME,
+	PY_ARG_TOTAL_TIME,
+	PY_ARG_ENTRY_ADDR,
+	PY_ARG_RET_ADDR,
+	PY_ARG_SYMNAME,
+	PY_ARG_RETVAL,
+};
+
+/* The order has to be aligned with enum py_args above. */
+static const char *py_args_table[] = {
+	"tid",
+	"depth",
+	"start_time",
+	"end_time",
+	"total_time",
+	"entry_addr",
+	"ret_addr",
+	"symname",
+	"retval",
+};
+
+#define INIT_PY_API_FUNC(func) \
+	do { \
+		__##func = dlsym(python_handle, #func); \
+		if (!__##func) { \
+			pr_err("dlsym for \"" #func "\" is failed!\n"); \
+			return -1; \
+		} \
+	} while (0)
+
+static char *remove_py_suffix(char *py_name)
+{
+	char *ext = strrchr(py_name, '.');
+
+	if (!ext)
+		return NULL;
+
+	*ext = '\0';
+	return py_name;
+}
+
+/* Import python module that is given by -p option */
+static int import_python_module(char *py_pathname)
+{
+	char py_sysdir[PATH_MAX];
+	if (absolute_dirname(py_pathname, py_sysdir) == NULL)
+		return -1;
+
+	/* Set path to import a python module. */
+	__PySys_SetPath(py_sysdir);
+	pr_dbg("PySys_SetPath(\"%s\") is done!\n", py_sysdir);
+
+	char *py_basename = basename(py_pathname);
+	remove_py_suffix(py_basename);
+
+	pName = __PyString_FromString(py_basename);
+	pModule = __PyImport_Import(pName);
+	if (pModule == NULL) {
+		__PyErr_Print();
+		pr_warn("%s.py cannot be imported!\n", py_pathname);
+		return -1;
+	}
+
+	return 0;
+}
+
+#ifdef NOT_LIBMCOUNT
+int python_uftrace_data_entry(struct ftrace_task_handle *task,
+			      struct uftrace_record *rstack)
+{
+	if (unlikely(!pFuncEntry))
+		return -1;
+
+	int tid = task->tid;
+	int depth = rstack->depth;
+	uint64_t start_time = rstack->time;
+	unsigned long entry_addr = rstack->addr;
+
+	struct sym *sym = find_symtabs(&task->h->sessions.first->symtabs, entry_addr);
+	char *symname = symbol_getname(sym, entry_addr);
```

Comment:
It's important to call `symbol_putname()` after using the symbol name to avoid potential memory leaks.